cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java

```diff
@@ -0,0 +1,93 @@
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.reflect.TypeToken;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Stage to write system metadata for the application.
+ */
+public class SystemMetadataWriterStage extends AbstractStage<ApplicationWithPrograms> {
+
+  private final MetadataStore metadataStore;
+
+  public SystemMetadataWriterStage(MetadataStore metadataStore) {
+    super(TypeToken.of(ApplicationWithPrograms.class));
+    this.metadataStore = metadataStore;
+  }
+
+  @Override
+  public void process(ApplicationWithPrograms input) throws Exception {
+    // add system metadata for apps
+    ApplicationSpecification appSpec = input.getSpecification();
+    AbstractSystemMetadataWriter appSystemMetadataWriter = new AppSystemMetadataWriter(metadataStore, input.getId(),
+                                                                                       appSpec);
+    appSystemMetadataWriter.write();
+
+    // add system metadata for programs
+    Id.Program programId;
+    List<AbstractSystemMetadataWriter> programSystemMetadataWriters = new ArrayList<>();
+    for (MapReduceSpecification mrSpec : appSpec.getMapReduce().values()) {
+      programId = Id.Program.from(input.getId(), ProgramType.MAPREDUCE, mrSpec.getName());
+      programSystemMetadataWriters.add(new ProgramSystemMetadataWriter(metadataStore, programId, mrSpec));
+    }
+    for (FlowSpecification flowSpec : appSpec.getFlows().values()) {
+      programId = Id.Program.from(input.getId(), ProgramType.FLOW, flowSpec.getName());
+      programSystemMetadataWriters.add(new ProgramSystemMetadataWriter(metadataStore, programId, flowSpec));
+    }
+    for (WorkflowSpecification workflowSpec : appSpec.getWorkflows().values()) {
+      programId = Id.Program.from(input.getId(), ProgramType.WORKFLOW, workflowSpec.getName());
+      programSystemMetadataWriters.add(new ProgramSystemMetadataWriter(metadataStore, programId, workflowSpec));
+    }
+    for (ServiceSpecification serviceSpec : appSpec.getServices().values()) {
+      programId = Id.Program.from(input.getId(), ProgramType.SERVICE, serviceSpec.getName());
+      programSystemMetadataWriters.add(new ProgramSystemMetadataWriter(metadataStore, programId, serviceSpec));
+    }
+    for (SparkSpecification sparkSpec : appSpec.getSpark().values()) {
+      programId = Id.Program.from(input.getId(), ProgramType.SPARK, sparkSpec.getName());
+      programSystemMetadataWriters.add(new ProgramSystemMetadataWriter(metadataStore, programId, sparkSpec));
+    }
+    for (WorkerSpecification workerSpec : appSpec.getWorkers().values()) {
+      programId = Id.Program.from(input.getId(), ProgramType.WORKER, workerSpec.getName());
+      programSystemMetadataWriters.add(new ProgramSystemMetadataWriter(metadataStore, programId, workerSpec));
+    }
+
+    for (AbstractSystemMetadataWriter programSystemMetadataWriter : programSystemMetadataWriters) {
```

Comment:
Why need to add all writers to a list and then loop it again to write? Can we just write in the for-loop above?